master/internal/stream/system_intg_test.go

```diff
@@ -0,0 +1,260 @@
+	require.Equal(t, "test", dataStr)
+}
+
+func TestStartup(t *testing.T) {
+	ctx := context.TODO()
+	pgDB, cleanup := db.MustResolveNewPostgresDatabase(t)
+	defer func() {
+		fmt.Println("cleaning up?")
+		cleanup()
+	}()
+
+	fmt.Println("START OF TEST!")
+	trials := streamdata.GenerateStreamTrials()
+	trials.MustMigrate(t, pgDB, "file://../../static/migrations")
+
+	startupMessage := StartupMsg{
+		Known: KnownKeySet{
+			Trials: "1,2,3",
+		},
+		Subscribe: SubscriptionSpecSet{
+			Trials: &TrialSubscriptionSpec{
+				ExperimentIds: []int{1}, // trials 1,2,3
+				Since:         0,
+			},
+		},
+	}
+
+	tester := startupReadWriter{
+		startupMessage: &startupMessage,
+	}
+	publisherSet := NewPublisherSet()
+	err := publisherSet.entrypoint(ctx, &tester, simpleUpsert, recordDeletion)
+	require.NoError(t, err)
+
+	deletions, trialMsgs, err := splitDeletionsAndTrials(tester.data)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(deletions), "did not receive 1 deletion message")
+	require.Equal(t, "0", deletions[0], "expected deleted trials to be 0, not %s", deletions[0])
+	require.Equal(t, 0, len(trialMsgs), "received unexpected trial message")
+
+	// don't know about trial 3, and trial 4 doesn't exist
+	startupMessage = StartupMsg{
+		Known: KnownKeySet{
+			Trials: "1,2,4",
+		},
+		Subscribe: SubscriptionSpecSet{
+			Trials: &TrialSubscriptionSpec{
+				ExperimentIds: []int{1}, // trials 1,2,3
+				Since:         0,
+			},
+		},
+	}
+	tester.startupMessage = &startupMessage
+	publisherSet = NewPublisherSet()
+	err = publisherSet.entrypoint(ctx, &tester, simpleUpsert, recordDeletion)
+	require.NoError(t, err)
+	deletions, trialMsgs, err = splitDeletionsAndTrials(tester.data)
+	require.NoError(t, err)
+	fmt.Println("deletions:", deletions)
+	require.Equal(t, 1, len(deletions), "did not receive 1 deletion message")
+	require.Equal(t, "4", deletions[0], "expected deleted trials to be 4, not %s", deletions[0])
+	require.Equal(t, 1, len(trialMsgs), "received unexpected trial message")
+	require.Equal(t, 3, trialMsgs[0].ID, "expected trialMsg with ID 3, received ID %d",
+		trialMsgs[0].ID)
+
+	fmt.Println("END OF TEST!")
+
+	// Subscribe to all known trials, but 4 doesn't exist
+	//startupMessage = StartupMsg{
+	//	Known: KnownKeySet{
+	//		Trials: "1,2,3,4",
+	//	},
+	//	Subscribe: SubscriptionSpecSet{
+	//		Trials: &TrialSubscriptionSpec{
+	//			TrialIds: []int{1, 2, 3, 4},
+	//			Since:    0,
+	//		},
+	//	},
+	//}
+	//tester.startupMessage = &startupMessage
+	//err = publisherSet.entrypoint(ctx, &tester, simpleUpsert, recordDeletion)
+	//require.NoError(t, err)
+	//deletions, trialMsgs, err = splitDeletionsAndTrials(tester.data)
+	//require.NoError(t, err)
+	//require.Equal(t, 1, len(deletions), "did not receive 1 deletion message")
+	//require.Equal(t, "4", deletions[0], "expected deleted trials to be 4, not %s", deletions[0])
+	//require.Equal(t, 0, len(trialMsgs), "received unexpected trial message")
+}
+
+func TestStartupTrial(t *testing.T) {
```

Comment:
These tests look pretty solid right now.

I'd like to see some streaming tests too, probably in a way that modifies the database.  We'll have some fairly complex SQL triggers that I think we'll want test coverage for.

Also, crazy idea: what if you made these tests json-based?  Like your expected values can just be strings and you can just string match them, no need to mess with the splitDeletionsAndTrials or anything like that.  I could see a future where the only thing you need to do to add tests for a new streaming type is write the initial database, a set of startup messages with expected startup outputs, and a set of database changes to apply with expected streaming outputs.